src/compiler/program.ts

```diff
@@ -386,6 +393,117 @@ namespace ts {
         allDiagnostics?: Diagnostic[];
     }
 
+    export function isProgramUptoDate(program: Program, rootFileNames: string[], newOptions: CompilerOptions,
+        getSourceVersion: (path: Path) => string, fileExists: (fileName: string) => boolean, hasInvalidatedResolution: HasInvalidatedResolution): boolean {
+        // If we haven't create a program yet, then it is not up-to-date
+        if (!program) {
+            return false;
+        }
+
+        // If number of files in the program do not match, it is not up-to-date
+        if (program.getRootFileNames().length !== rootFileNames.length) {
+            return false;
+        }
+
+        // If any file is not up-to-date, then the whole program is not up-to-date
+        for (const file of program.getSourceFiles()) {
+            if (!sourceFileUpToDate(program.getSourceFile(file.fileName))) {
+                return false;
+            }
+        }
+
+        // If any of the missing file paths are now created
+        if (program.getMissingFilePaths().some(missingFilePath => fileExists(missingFilePath))) {
+            return false;
+        }
+
+        const currentOptions = program.getCompilerOptions();
+        // If the compilation settings do no match, then the program is not up-to-date
+        if (!compareDataObjects(currentOptions, newOptions)) {

```

Comment:
I would keep this way since I have just moved from services to the compiler utilities so that we can unify the logic. If there is need to do more in this function, we could handle that on demand later.